File: AVS3/triangle.py
from dataclasses import dataclass

from .color import Color
from .point import Point


@dataclass
class Triangle:
    '''
    Хранит информацию о трегуольнике.
    '''

    color: Color
    a: Point
    b: Point
    c: Point

    @staticmethod
    def read(buffer) -> 'Triangle':
        color = Color.read(buffer)
        buffer.skip_whitespaces()
        a = Point.read(buffer)
        buffer.skip_whitespaces()
        b = Point.read(buffer)
        buffer.skip_whitespaces()
        c = Point.read(buffer)
        return Triangle(color, a, b, c)

    def perimiter(self):
        return (self.a.distance(self.b)
                + self.b.distance(self.c)
                + self.c.distance(self.a))

    def __str__(self) -> str:
        return (f'Triangle: color={self.color}, '
                f'points={self.a} - {self.b} - {self.c}')

File: lua/triangle.lua
local Color = require "color"
local Point = require "point"

local Triangle = {}
Triangle.__index = Triangle

function Triangle:read(buffer)
	result = setmetatable({}, self)
	result.color = Color:read(buffer)
	result.a = Point:read(buffer)
	result.b = Point:read(buffer)
	result.c = Point:read(buffer)
	setmetatable(result, Triangle)
	return result
end

function Triangle:show()
	return string.format("Triangle: color=%s, points=(%d, %d)-(%d, %d)-(%d, %d)",
	                     self.color, self.a.x, self.a.y,
			                 self.b.x, self.b.y,
			                 self.c.x, self.c.y)
end

function Triangle:perimiter()
	return self.a:distance(self.b) + self.b:distance(self.c) + self.c:distance(self.a)
end

return Triangle
